### PR TITLE
Concurrent Modification Exception PLAT-2666

### PR DESCRIPTION
**Pull request by EladCohenKaltura**
PLAT-2666
Iterating over collections while deleting entries from them
Created Concurrent Modification Exception
Using iterator to avoid exception

### CODE REVIEW

File: KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java [hunk 1]
```diff
@@ -118,8 +120,16 @@ public void load(final List<String> entryIds, boolean periodic) {
 						}
 						
 						// create sync-point 30 seconds before the trigger time
-						date.setTime((cuePoint.triggeredAt / 1000) - 30000);
+						long scheduledTime = (cuePoint.triggeredAt / 1000) - 30000;
```
Comment (tan-tan-kanarek) [inline KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java]: This test is not needed as the triggeredAt is always timestamp larger than 1980 which is a large number.

File: KalturaMediaServer/src/com/kaltura/media/server/api/KalturaWebServicesServer.java [hunk 2]
```diff
@@ -42,11 +44,12 @@ public Endpoint addService(IWebService service)
 	}
 
 	public void shutdown() {
-		Set<String> keys = services.keySet();
-		for(String key : keys)
+		Iterator<Map.Entry<String, Endpoint>> serviceIter = services.entrySet().iterator();
```
Comment (tan-tan-kanarek) [inline KalturaMediaServer/src/com/kaltura/media/server/api/KalturaWebServicesServer.java]: This set never modified, and definitely not in shutting down.

File: KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java [hunk 3]
```diff
@@ -241,48 +253,40 @@ protected void onCuePoint(KalturaLiveEntry entry, KalturaCuePoint cuePoint) {
 	protected void onUnPublish(String entryId) {
 		
 		synchronized (cuePointsCreators) {
-			CuePointsCreator removeCreator = null;
-			
-			for(CuePointsCreator cuePointsCreator: cuePointsCreators.values()){
+			Iterator<Map.Entry<Integer, CuePointsCreator>> cuePointsCreatorIter = cuePointsCreators.entrySet().iterator();
```
Comment (tan-tan-kanarek) [inline KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java]: This is inside synchronized  block, so it's not needed.

File: KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java [hunk 4]
```diff
@@ -241,48 +253,40 @@ protected void onCuePoint(KalturaLiveEntry entry, KalturaCuePoint cuePoint) {
...
-			if(removeCreator != null) {
-				cuePointsCreators.remove(removeCreator.interval);
-			}
 		}
 		
 		synchronized (cuePointsLoaders) {
-			CuePointsLoader removeLoader = null;
-			
-			for(CuePointsLoader cuePointsLoader: cuePointsLoaders){
+			Iterator<CuePointsLoader> cuePointsLoaderIter = cuePointsLoaders.iterator();
```
Comment (tan-tan-kanarek) [inline KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java]: This is inside synchronized block, so it's not needed.

File: KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java [hunk 5]
```diff
@@ -290,23 +294,31 @@ protected void onUnPublish(String entryId) {
...
 	@Override
 	public void stop() {
-		synchronized (cuePointsCreators) {
-			for(CuePointsCreator cuePointsCreator: cuePointsCreators.values()){
-				synchronized (cuePointsCreator) {
-					cuePointsCreator.timer.cancel();
-					cuePointsCreator.timer.purge();
+		try {
+			synchronized (cuePointsCreators) {
+				Iterator<Map.Entry<Integer, CuePointsCreator>> cuePointsCreatorIter = cuePointsCreators.entrySet().iterator();
```
Comment (tan-tan-kanarek) [inline KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java]: This is inside synchronized block, so it's not needed.

File: KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java (hunk 1 above)
Comment (EladCohenKaltura) [inline KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java]: This was inserted due to cases where the triggered at was small and created a situation that the Timer failed, consider the fact that the user can set values through API calls

File: KalturaMediaServer/src/com/kaltura/media/server/api/KalturaWebServicesServer.java (hunk 2 above)
Comment (EladCohenKaltura) [inline KalturaMediaServer/src/com/kaltura/media/server/api/KalturaWebServicesServer.java]: In the next section we can see the services.remove(key); which removes the entry , did not get what you ment

File: KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java (hunk 3 above)
Comment (EladCohenKaltura) [inline KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java]: the problem can occur on a single thread

File: KalturaMediaServer/src/com/kaltura/media/server/api/KalturaWebServicesServer.java (hunk 2 above)
Comment (tan-tan-kanarek) [inline KalturaMediaServer/src/com/kaltura/media/server/api/KalturaWebServicesServer.java]: This is the only place that we remove from this collection, and therefore there is no need to synchronize it. But I don't care enough, you can keep this code.

File: KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java (hunk 1 above)
Comment (tan-tan-kanarek) [inline KalturaMediaServer/src/com/kaltura/media/server/managers/KalturaCuePointsManager.java]: If you set value that is smaller than 1980 using the API, the server will translate it to relative time and will add the specified number to time(), that's why this number can't be small.
If you know of a specific case that it was small, then it's a bug, and we should solve it, offline.

**Pull request merged**